pkg/arm/arm.go

```diff
@@ -77,10 +83,133 @@ func (g *simpleGenerator) Generate(ctx context.Context, cs *api.OpenShiftManaged
 		return nil, err
 	}
 
-	var azuretemplate map[string]interface{}
-	err = json.Unmarshal(azuredeploy, &azuretemplate)
-	if err != nil {
+	var original, copied map[string]interface{}
+	if err := json.Unmarshal(azuredeploy, &original); err != nil {
+		return nil, err
+	}
+	if err := json.Unmarshal(azuredeploy, &copied); err != nil {
+		return nil, err
+	}
+	if err := hashScaleSets(original, copied); err != nil {
 		return nil, err
 	}
-	return azuretemplate, nil
+	return original, nil
+}
+
+func hashScaleSets(original, copied map[string]interface{}) error {
+	for key, value := range copied {
+		if key != "resources" {
+			continue
+		}
+
+		for _, r := range value.([]interface{}) {
+			resource, ok := r.(map[string]interface{})
+			if !ok {
+				continue
+			}
+
+			if !isScaleSet(resource) {
+				continue
+			}
+
+			// cleanup capacity
+			deleteField(resource, "sku", "capacity")
+
+			// cleanup previous hash
+			deleteField(resource, "tags", hashKey)
+
+			// hash scale set
+			data, err := json.Marshal(resource)
+			if err != nil {
+				return err
+			}
+			hf := sha256.New()
+			fmt.Fprintf(hf, string(data))
+			h := base64.StdEncoding.EncodeToString(hf.Sum(nil))
+
+			// update tags in the original template
+			role := getRole(resource)
+			if added := addTag(role, hashKey, h, original); !added {
+				return fmt.Errorf("could not tag ARM template with new hash for role %q", role)
+			}
+		}
+	}
+	return nil
+}
+
+func isScaleSet(resource map[string]interface{}) bool {
+	for k, v := range resource {
+		if k == "type" && v.(string) == "Microsoft.Compute/virtualMachineScaleSets" {
+			return true
+		}
+	}
+	return false
+}
+
+func deleteField(resource map[string]interface{}, parent, field string) {
+	for key, value := range resource {
+		if key != parent {
+			continue
+		}
+		if p, ok := value.(map[string]interface{}); ok {
+			delete(p, field)
+			resource[key] = p
+			return
+		}
+	}
+}
+
+func getRole(resource map[string]interface{}) string {
+	for k, v := range resource {
+		if k == "name" && strings.HasPrefix(v.(string), "ss-") {
+			return v.(string)[3:]
+		}
+	}
+	return ""
+}
+
+func addTag(role, key, value string, azuretemplate map[string]interface{}) bool {
+	var added bool
+
+out:
+	for k, resources := range azuretemplate {
+		if k != "resources" {
+			continue
+		}
+
+		for _, r := range resources.([]interface{}) {
+			resource, ok := r.(map[string]interface{})
+			if !ok {
+				continue
+			}
+
+			if !isScaleSet(resource) {
+				continue
+			}
+			if !hasRole(role, resource) {
+				continue
+			}
+			newTags := make(map[string]interface{})
+			if tags, ok := resource["tags"].(map[string]interface{}); ok {
+				for k, v := range tags {
+					newTags[k] = v
+				}
+			}
+			newTags[key] = value
+			resource["tags"] = newTags
+			added = true
+			break out
```

Comment:
these break labels are really not intuitive, I had to go read https://golang.org/ref/spec#Break_statements to figure out it goes to the next statement out of the loop - weird.

Reply:
Agree :)

Reply:
This can also be a return, will update